File: androidApp/src/main/java/com/luciddreaming/android/viewmodel/SettingsViewModel.kt
package com.luciddreaming.android.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.luciddreaming.shared.data.preferences.AndroidConnectionPreferences
import com.luciddreaming.shared.data.preferences.AndroidSettingsPreferences
import com.luciddreaming.shared.data.preferences.AppSettings
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

class SettingsViewModel(application: Application) : AndroidViewModel(application) {
    private val settingsPreferences = AndroidSettingsPreferences(application)
    private val connectionPreferences = AndroidConnectionPreferences(application)

    private val _settings = MutableStateFlow(AppSettings())
    val settings: StateFlow<AppSettings> = _settings.asStateFlow()

    init {
        loadSettings()
    }

    private fun loadSettings() {
        viewModelScope.launch {
            settingsPreferences.settings.collect {
                _settings.value = it
            }
        }
    }

    fun updateRefreshInterval(interval: Int) {
        viewModelScope.launch {
            val currentSettings = settingsPreferences.settings.first()
            settingsPreferences.updateSettings(
                refreshInterval = interval,
                autoRefresh = currentSettings.autoRefresh,
                darkMode = currentSettings.darkMode
            )
        }
    }

    fun updateAutoRefresh(enabled: Boolean) {
        viewModelScope.launch {
            val currentSettings = settingsPreferences.settings.first()
            settingsPreferences.updateSettings(
                refreshInterval = currentSettings.refreshInterval,
                autoRefresh = enabled,
                darkMode = currentSettings.darkMode
            )
        }
    }

    fun updateDarkMode(enabled: Boolean) {
        viewModelScope.launch {
            val currentSettings = settingsPreferences.settings.first()
            settingsPreferences.updateSettings(
                refreshInterval = currentSettings.refreshInterval,
                autoRefresh = currentSettings.autoRefresh,
                darkMode = enabled
            )
        }
    }

    fun resetSettings() {
        viewModelScope.launch {
            settingsPreferences.resetSettings()
        }
    }

    fun disconnect() {
        viewModelScope.launch {
            connectionPreferences.clearConnectionSettings()
        }
    }
}

File: iosApp/LucidDreaming/ViewModels/SettingsViewModel.swift
import Foundation
import Combine
import Shared

class SettingsViewModel: ObservableObject {
    @Published var settings: AppSettings = AppSettings()
    
    private let settingsPreferences = iOSSsettingsPreferences()
    private let connectionPreferences = iOSConnectionPreferences()
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        loadSettings()
    }
    
    private func loadSettings() {
        settingsPreferences.settings
            .sink {
                self.settings = $0
            }
            .store(in: &cancellables)
    }
    
    func updateRefreshInterval(interval: Int) {
        Task {
            await settingsPreferences.updateSettings(
                refreshInterval: interval,
                autoRefresh: settings.autoRefresh,
                darkMode: settings.darkMode
            )
        }
    }
    
    func updateAutoRefresh(enabled: Bool) {
        Task {
            await settingsPreferences.updateSettings(
                refreshInterval: settings.refreshInterval,
                autoRefresh: enabled,
                darkMode: settings.darkMode
            )
        }
    }
    
    func updateDarkMode(enabled: Bool) {
        Task {
            await settingsPreferences.updateSettings(
                refreshInterval: settings.refreshInterval,
                autoRefresh: settings.autoRefresh,
                darkMode: enabled
            )
        }
    }
    
    func resetSettings() {
        Task {
            await settingsPreferences.resetSettings()
        }
    }
    
    func disconnect() {
        Task {
            await connectionPreferences.clearConnectionSettings()
        }
    }
}
